src/platform/datapath_raw_xdp.c

```diff
@@ -428,40 +428,108 @@ CxPlatDpRawGetDapathSize(
+
+        if (Queue->TxXsk != NULL) {
+            QUIC_STATUS Status;
+            XSK_STATISTICS Stats;
+            uint32_t StatsSize = sizeof(Stats);
+            Status = XskGetSockopt(Queue->TxXsk, XSK_SOCKOPT_STATISTICS, &Stats, &StatsSize);
+            if (QUIC_SUCCEEDED(Status)) {
+                printf("[%u-%u]txInvalidDescriptors: %llu\n", Interface->IfIndex, i, Stats.txInvalidDescriptors);
+            }
+            CloseHandle(Queue->TxXsk);
+        }
+
+        if (Queue->TxBuffers != NULL) {
+            CxPlatFree(Queue->TxBuffers, TX_BUFFER_TAG);
+        }
+
+        if (Queue->RxProgram != NULL) {
+            CloseHandle(Queue->RxProgram);
+        }
+
+        if (Queue->RxXsk != NULL) {
+            QUIC_STATUS Status;
+            XSK_STATISTICS Stats;
+            uint32_t StatsSize = sizeof(Stats);
+            Status = XskGetSockopt(Queue->RxXsk, XSK_SOCKOPT_STATISTICS, &Stats, &StatsSize);
+            if (QUIC_SUCCEEDED(Status)) {
+                printf("[%u-%u]rxDropped: %llu\n", Interface->IfIndex, i, Stats.rxDropped);
+                printf("[%u-%u]rxInvalidDescriptors: %llu\n", Interface->IfIndex, i, Stats.rxInvalidDescriptors);
+            }
+            CloseHandle(Queue->RxXsk);
+        }
+
+        if (Queue->RxBuffers != NULL) {
+            CxPlatFree(Queue->RxBuffers, RX_BUFFER_TAG);
+        }
+
+        CxPlatLockUninitialize(&Queue->TxLock);
+    }
+
+    if (Interface->Queues != NULL) {
+        CxPlatFree(Interface->Queues, QUEUE_TAG);
+    }
+
+    #pragma warning(pop)
+}
+
 _IRQL_requires_max_(PASSIVE_LEVEL)
 QUIC_STATUS
-CxPlatDpRawInitialize(
-    _Inout_ CXPLAT_DATAPATH* Datapath,
+CxPlatDpRawInterfaceInitialize(
+    _In_ XDP_DATAPATH* Xdp,
+    _Out_ XDP_INTERFACE* Interface,
+    _In_ uint32_t IfIndex,
     _In_ uint32_t ClientRecvContextLength
     )
 {
-    XDP_DATAPATH* Xdp = (XDP_DATAPATH*)Datapath;
-    CXPLAT_THREAD_CONFIG Config = {
-        0, 0, "XdpDatapathWorker", CxPlatXdpWorkerThread, Xdp
-    };
     const uint32_t RxHeadroom = sizeof(XDP_RX_PACKET) + ALIGN_UP(ClientRecvContextLength, uint32_t);
     const uint32_t RxPacketSize = ALIGN_UP(RxHeadroom + MAX_ETH_FRAME_SIZE, XDP_RX_PACKET);
     QUIC_STATUS Status;
 
-    CxPlatXdpReadConfig(Xdp);
-    Datapath->Cpu = Xdp->DatapathCpuNumber;
-    CxPlatDpRawGenerateCpuTable(Datapath);
+    Interface->IfIndex = IfIndex;
+    Interface->OffloadStatus.Receive.NetworkLayerXsum = Xdp->SkipXsum;
+    Interface->OffloadStatus.Receive.TransportLayerXsum = Xdp->SkipXsum;
+    Interface->OffloadStatus.Transmit.NetworkLayerXsum = Xdp->SkipXsum;
+    Interface->OffloadStatus.Transmit.NetworkLayerXsum = Xdp->SkipXsum;
 
-    Xdp->Queues = CxPlatAlloc(Xdp->QueueCount * sizeof(*Xdp->Queues), QUEUE_TAG);
-    if (Xdp->Queues == NULL) {
+    Status = CxPlatGetInterfaceRssQueueCount(Interface->IfIndex, &Interface->QueueCount);
+    if (QUIC_FAILED(Status)) {
         QuicTraceEvent(
             AllocFailure,
```

Comment:
1) It's the wrong error message type. 2) You don't need to log because the function itself already logs the point of failure.